src/gui/window.cpp

```diff
@@ -0,0 +1,284 @@
+	{
+		if (index < m_style_strs.size()) {
+			return &m_style_strs[index];
+		}
+		return nullptr;
+	}
+
+	void Window::reset()
+	{
+		m_elems.clear();
+		m_ordered_elems.clear();
+
+		m_root_elem = nullptr;
+
+		m_style_strs.clear();
+	}
+
+	void Window::read(std::istream &is, bool opening)
+	{
+		std::unordered_map<Elem *, std::string> elem_contents;
+		readElems(is, elem_contents);
+
+		readRootElem(is);
+		readStyles(is);
+
+		if (opening)
+			m_type = toWindowType(readU8(is));
+
+		// Assuming no earlier step failed, we can proceed to read in all the
+		// properties. Otherwise, reset the window entirely.
+		if (m_root_elem != nullptr) {
+			updateElems(elem_contents);
+		} else {
+			reset();
+		}
+	}
+
+	float Window::getPixelSize() const
+	{
+		return g_manager.getPixelSize(m_type);
+	}
+
+	d2f32 Window::getScreenSize() const
+	{
+		return g_manager.getScreenSize(m_type);
+	}
+
+	void Window::drawAll()
+	{
+		if (m_root_elem == nullptr) {
+			return;
+		}
+
+		rf32 parent_rect(getScreenSize());
+		m_root_elem->layout(parent_rect, parent_rect);
+
+		Canvas canvas(Texture::screen, getPixelSize());
+		m_root_elem->drawAll(canvas);
+	}
+
+	void Window::readElems(std::istream &is,
+			std::unordered_map<Elem *, std::string> &elem_contents)
+	{
+		// Read in all the new elements and updates to existing elements.
+		u32 num_elems = readU32(is);
+
+		std::unordered_map<std::string, std::unique_ptr<Elem>> new_elems;
+
+		for (size_t i = 0; i < num_elems; i++) {
+			u32 type = readU8(is);
+			std::string id = readNullStr(is);
+
+			// Make sure that elements have valid IDs. If the string has non-ID
+			// characters in it, though, we don't particularly care.
+			if (id.empty()) {
+				errorstream << "Element has empty ID" << std::endl;
+				continue;
+			}
+
+			// Each element has a size prefix stating how big the element is.
+			// This allows new fields to be added to elements without breaking
+			// compatibility. So, read it in as a string and save it for later.
+			std::string contents = readStr32(is);
+
+			// If this is a duplicate element, skip it right away.
+			if (new_elems.find(id) != new_elems.end()) {
+				errorstream << "Duplicate element \"" << id << "\"" << std::endl;
+				continue;
+			}
+
+			/* Now we need to decide whether to create a new element or to
+			 * modify the state of an already existing one. This allows
+			 * changing attributes of an element (like the style or the
+			 * element's children) while leaving leaving persistent state
+			 * intact (such as the position of a scrollbar or the contents of a
+			 * text field).
+			 */
+			std::unique_ptr<Elem> elem = nullptr;
+
+			// Search for a pre-existing element.
+			auto it = m_elems.find(id);
+
+			if (it == m_elems.end() || it->second->getType() != type) {
+				// If the element was not found or the existing element has the
+				// wrong type, create a new element.
+				elem = Elem::create((Elem::Type)type, *this, id);
```

Comment:
If an element with the same ID and a different type exists and we still add create the new element, doesn't this mean we now have two elements with the same ID and different types?

Reply:
No.  We're busy populating an empty `unordered_map`, so we either create a new element (possibly discarding the old one) or steal an existing element from the existing `unordered_map` in the `else` branch.  After we finish populating the new `unordered_map`, any remaining elements in the old one will be destroyed.